web/modules/custom/joinup_core/joinup_core.post_update.php

```diff
@@ -48,3 +51,48 @@ function joinup_core_post_update_0106400(): void {
   $index->reindex();
   $index->indexItems(-1, 'entity:taxonomy_term');
 }
+
+/**
+ * Migrate data about pinned entities into meta entities.
+ */
+function joinup_core_post_update_0106401(): void {
+  $entity_type_manager = \Drupal::entityTypeManager();
+
+  $get_storage = function (string $entity_type_id) use ($entity_type_manager) {
+    return $entity_type_manager->getStorage($entity_type_id);
+  };
+
+  $entity_ids = [];
+
+  // Retrieve community content to migrate.
+  $entity_ids['node'] = $get_storage('node')
+    ->getQuery()
+    ->condition('type', CommunityContentHelper::BUNDLES, 'IN')
+    ->condition('sticky', TRUE)
+    ->execute();
+
+  // Retrieve solutions to migrate.
+  $entity_ids['rdf_entity'] = $get_storage('rdf_entity')
+    ->getQuery()
+    ->condition('rid', 'solution')
+    ->exists('field_is_pinned_in')
+    ->execute();
+
+  foreach ($entity_ids as $entity_type_id => $ids) {
+    $storage = $get_storage($entity_type_id);
+    foreach ($storage->loadMultiple($ids) as $entity) {
+      if ($entity instanceof PinnableGroupContentInterface) {
+        if ($entity instanceof CommunityContentInterface) {
+          $entity->pin();
+        }
+        else {
+          /** @var \Drupal\joinup_group\Entity\GroupInterface[] $pinned_groups */
+          $pinned_groups = $entity->get('field_is_pinned_in')->referencedEntities();
```

Comment:
I did it like this because we should not use the Drupal API in an update hook, only in a post-update hook. But I see you used some trickery to store the values in state and then pick them up in the post-update. This is not very conventional but it might work :D